scrapy/core/engine.py

```diff
@@ -102,188 +109,175 @@ def close(self):
             # Will also close downloader
             return self._close_all_spiders()
         else:
-            return defer.succeed(self.downloader.close())
+            return defer_succeed(self.downloader.close())
 
-    def pause(self):
-        """Pause the execution engine"""
+    def pause(self) -> None:
         self.paused = True
 
-    def unpause(self):
-        """Resume the execution engine"""
+    def unpause(self) -> None:
         self.paused = False
 
-    def _next_request(self, spider):
-        slot = self.slot
-        if not slot:
-            return
+    def _next_request(self, spider: Spider) -> None:
+        if self.slot is None:
+            return None
 
         if self.paused:
-            return
+            return None
 
-        while not self._needs_backout(spider):
-            if not self._next_request_from_scheduler(spider):
-                break
+        while not self._needs_backout(spider) and self._next_request_from_scheduler(spider) is not None:
+            pass
 
-        if slot.start_requests and not self._needs_backout(spider):
+        if self.slot.start_requests is not None and not self._needs_backout(spider):
             try:
-                request = next(slot.start_requests)
+                request = next(self.slot.start_requests)
             except StopIteration:
-                slot.start_requests = None
+                self.slot.start_requests = None
             except Exception:
-                slot.start_requests = None
-                logger.error('Error while obtaining start requests',
-                             exc_info=True, extra={'spider': spider})
+                self.slot.start_requests = None
+                logger.error('Error while obtaining start requests', exc_info=True, extra={'spider': spider})
             else:
                 self.crawl(request, spider)
 
-        if self.spider_is_idle(spider) and slot.close_if_idle:
+        if self.spider_is_idle(spider) and self.slot.close_if_idle:
             self._spider_idle(spider)
 
-    def _needs_backout(self, spider):
-        slot = self.slot
+    def _needs_backout(self, spider: Spider) -> bool:
         return (
             not self.running
-            or slot.closing
+            or self.slot is not None and self.slot.closing
```

Comment:
This gives the impression that `self.slot` may be `None`, which does not seem to be the case, given the implementation of `_next_request` and assuming that is the only caller of `_needs_backout`.

Would an `assert` before the return be enough for mypy?

Same applies to `_next_request_from_scheduler` below.